### PR TITLE
update (passive_anticheat): unlocked LUA macro blocker\conf

### PR DESCRIPTION
**Pull request by acidmanifesto**
**Changes proposed**:

- Implement unlocked LUA macro Blocker
- New Conf Anticheat.LUAblocker = 0/1
- Implement new prepared statement
- Add new tables to the character and world db
- Add ip addresses to logging

Following moved to https://github.com/TrinityCore/TrinityCoreCustomChanges/pull/105 for a build fix with 3.3.5 base.
~~make adjustments due to https://github.com/TrinityCore/TrinityCore/commit/2a587e5ac43ffd274ee7a1feb3ce2e35c2ae2978  reported by @Aokromes with feed back from @Shauren and @jackpoz 
move a helper to correct a exploit issue.~~

There is currently over a hundred of homebrew LUA unlockers out there in the wild thanks to open source projects.
Alot of them however do not work due to a intentional thing in the core causing them to hang If i am seeing that right in debugger.

So instead of finding more ways to try to block a unlocker, lets find ways to block the macro if a unlocker actually goes thru successifully.

We add a check for macros and put them on a list of potentional cheaters in the characters if the macro is listed in the world db table and as enabled.

So the big issue with LUA unlockers is it gives the player access to restricted and protected functions of the client that is suppose to remain inaccessible to the player. The big abuse is the ability to cast spells regardless of cooldowns and requirements being met. This makes for a nasty issue with pvp and botting in most cases if the botting utilizes unlocked LUA macro usuage.

With the spell casting being the cause for a abuse thru lua locker being its prime usage, we limit it on the repeat which is where the abuse occurs, if the player is added in and the macro detection is enabled, which is should be in the db by default.

At the moment there is 1 new confs for this.
```
Anticheat.LUAblocker = 0/1
```

character db sql to be ran for current users of the anticheat:
```
CREATE TABLE IF NOT EXISTS `lua_cheaters` ( 
   `guid` int unsigned NOT NULL DEFAULT 0, 
   `account` int unsigned NOT NULL DEFAULT 0, 
   `macro` varchar(255) NOT NULL DEFAULT '', 
     PRIMARY KEY (`guid`,`account`) 
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```
world db sql to be ran from current users of the anticheat:
```
CREATE TABLE IF NOT EXISTS `lua_private_functions` ( 
   `function_name` varchar(255) NOT NULL DEFAULT '0', 
   `enabled` tinyint NOT NULL DEFAULT '1', 
   PRIMARY KEY (`function_name`) 
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4; 
  
 INSERT INTO `lua_private_functions` VALUES ('CastSpellByName', '1'); 
 INSERT INTO `lua_private_functions` VALUES ('RunMacroText', '1');
```

And for people wanting to cherry pick this, don't be a assclown and strip out the scam notice or the authoring.

**Target branch(es)**: 335-passive_anticheat

**Issues addressed**: Fixes #

**Tests performed**: (Does it build, tested in-game, etc)

**Known issues and TODO list**:

- [ ] 
- [ ]

### CODE REVIEW

File: sql/world.anticheat.sql [hunk 1]
```diff
@@ -26,3 +26,13 @@ INSERT INTO `command` (`name`, `help`) VALUES
 ('anticheat purge', 'Syntax: .anticheat purge\r\n\r\nDeletes stored statistics of daily_players_reports table.'),
 ('anticheat warn', 'Syntax: .anticheat warn [$charactername]\r\n\r\nSends individual player they are being monitored for possible cheating.'),
 ('anticheat handle', 'Syntax: .anticheat handle 0 or 1\r\n\r\n0 Turns off and 1 Turns On Anticheat');
+
+DROP TABLE IF EXISTS `lua_private_functions`;
+CREATE TABLE IF NOT EXISTS `lua_private_functions` (
+  `function_name` varchar(255) NOT NULL DEFAULT '0',
+  `enabled` tinyint(4) NOT NULL DEFAULT '1',
+  PRIMARY KEY (`function_name`,`enabled`)
```
Comment (jackpoz) [inline sql/world.anticheat.sql]: isn't function_name enough as key ?

File: sql/world.anticheat.sql [hunk 2]
```diff
@@ -26,3 +26,13 @@ INSERT INTO `command` (`name`, `help`) VALUES
 ('anticheat purge', 'Syntax: .anticheat purge\r\n\r\nDeletes stored statistics of daily_players_reports table.'),
 ('anticheat warn', 'Syntax: .anticheat warn [$charactername]\r\n\r\nSends individual player they are being monitored for possible cheating.'),
 ('anticheat handle', 'Syntax: .anticheat handle 0 or 1\r\n\r\n0 Turns off and 1 Turns On Anticheat');
+
+DROP TABLE IF EXISTS `lua_private_functions`;
+CREATE TABLE IF NOT EXISTS `lua_private_functions` (
+  `function_name` varchar(255) NOT NULL DEFAULT '0',
+  `enabled` tinyint(4) NOT NULL DEFAULT '1',
```
Comment (jackpoz) [inline sql/world.anticheat.sql]: Do you ever check if it's enabled or disabled ?  I think you can just drop this column

Comment (acidmanifesto) [reply sql/world.anticheat.sql]: currently i have it checking the enabled on the PrepareStatement
image

File: sql/world.anticheat.sql (hunk 1 above)
Comment (acidmanifesto) [reply sql/world.anticheat.sql]: yes. i do believe it is, i honestly do not see why not. i will make that adjustment now

File: sql/character.anticheat.sql [hunk 3]
```diff
@@ -45,6 +45,18 @@ CREATE TABLE IF NOT EXISTS `players_reports_status` (
   PRIMARY KEY (`guid`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
 
+-- Dumping structure for table characters.lua_cheaters
+DROP TABLE IF EXISTS `lua_cheaters`;
+CREATE TABLE IF NOT EXISTS `lua_cheaters` (
+  `guid` int(11) NOT NULL,
```
Comment (jackpoz) [inline sql/character.anticheat.sql]: Please use the same type as other tables, so
```sql
`guid` int unsigned NOT NULL DEFAULT 0,
```

Comment (acidmanifesto) [reply sql/character.anticheat.sql]: Ah. Thats what i forgot. I couldnt remember the past month what i had left to do. Ill get to it later.

File: sql/character.anticheat.sql [hunk 4]
```diff
@@ -45,6 +45,18 @@ CREATE TABLE IF NOT EXISTS `players_reports_status` (
...
+  `account` int(11) NOT NULL,
```
Comment (jackpoz) [inline sql/character.anticheat.sql]: From auth.account table, this should be `int unsigned NOT NULL`

Comment (jackpoz) [reply sql/character.anticheat.sql]: What do you need account for btw ? isn't guid unique already ? No other table has `account`

File: sql/character.anticheat.sql [hunk 5]
```diff
@@ -45,6 +45,18 @@ CREATE TABLE IF NOT EXISTS `players_reports_status` (
...
+  `macro` varchar(255) DEFAULT NULL,
+    PRIMARY KEY (`guid`,`account`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+
+-- Dumping data for table characters.lua_cheaters: ~0 rows (approximately)
+DELETE FROM `lua_cheaters`;
```
Comment (jackpoz) [inline sql/character.anticheat.sql]: Please remove this as we are creating the table from scratch

File: sql/character.anticheat.sql [hunk 6]
```diff
@@ -45,6 +45,18 @@ CREATE TABLE IF NOT EXISTS `players_reports_status` (
   PRIMARY KEY (`guid`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
 
+-- Dumping structure for table characters.lua_cheaters
+DROP TABLE IF EXISTS `lua_cheaters`;
```
Comment (jackpoz) [inline sql/character.anticheat.sql]: Please remove this so it errors out if someone tries to run it again, so they don't lose the data

File: src/server/game/Anticheat/AnticheatMgr.cpp [hunk 7]
```diff
@@ -90,6 +90,81 @@ AnticheatMgr::~AnticheatMgr()
     m_Players.clear();
 }
 
+void AnticheatMgr::LoadBlockedLuaFunctions()
+{
+    uint32 oldmsTime = getMSTime();
+    auto pstmt = WorldDatabase.GetPreparedStatement(WORLD_SEL_ANTICHEAT_FUNCTIONS);
+    auto result = WorldDatabase.Query(pstmt);
+    if (!sWorld->getBoolConfig(CONFIG_LUABLOCKER_ENABLE))
```
Comment (jackpoz) [inline src/server/game/Anticheat/AnticheatMgr.cpp:95]: please move this block before doing any DB query as they are not needed if it's disabled

File: src/server/game/Anticheat/AnticheatMgr.cpp [hunk 8]
```diff
@@ -90,6 +90,81 @@ AnticheatMgr::~AnticheatMgr()
...
+    for (auto& kv : _luaBlockedFunctions)
+    {
+        for (uint8 i = 0; i < NUM_ACCOUNT_DATA_TYPES; ++i)
+        {
+            std::string currentData = accountData[i].Data;
+            std::size_t pos = currentData.find(kv.first);
+            if (pos != std::string::npos)
+            {
+                const static std::size_t defaultLength = 200;
+                std::size_t minPos = int64(int(pos) - 50) < 0 ? 0 : pos - 50;
```
Comment (jackpoz) [inline src/server/game/Anticheat/AnticheatMgr.cpp:157]: What is this int64 and int conversion ? pos is size_t type, it should stay as size_t.
What is the logic about 200 and 50 ?

File: sql/character.anticheat.sql (hunk 3 above)
Comment (acidmanifesto) [reply sql/character.anticheat.sql]: https://github.com/TrinityCore/TrinityCoreCustomChanges/commit/1d1cdda4d4a033b99dde03f0de4606956be0f989

File: sql/character.anticheat.sql (hunk 4 above)
Comment (acidmanifesto) [reply sql/character.anticheat.sql]: Character table has account.  It soo we will know which character specifically used the hack for further evaluation

Comment (acidmanifesto) [reply sql/character.anticheat.sql]: https://github.com/TrinityCore/TrinityCoreCustomChanges/commit/1d1cdda4d4a033b99dde03f0de4606956be0f989

File: sql/character.anticheat.sql (hunk 5 above)
Comment (acidmanifesto) [reply sql/character.anticheat.sql]: https://github.com/TrinityCore/TrinityCoreCustomChanges/commit/1d1cdda4d4a033b99dde03f0de4606956be0f989

File: sql/character.anticheat.sql (hunk 6 above)
Comment (acidmanifesto) [reply sql/character.anticheat.sql]: https://github.com/TrinityCore/TrinityCoreCustomChanges/commit/1d1cdda4d4a033b99dde03f0de4606956be0f989

File: src/server/game/Anticheat/AnticheatMgr.cpp (hunk 7 above)
Comment (acidmanifesto) [reply src/server/game/Anticheat/AnticheatMgr.cpp:95]: https://github.com/TrinityCore/TrinityCoreCustomChanges/commit/1d1cdda4d4a033b99dde03f0de4606956be0f989

File: src/server/game/Anticheat/AnticheatMgr.cpp (hunk 8 above)
Comment (acidmanifesto) [reply src/server/game/Anticheat/AnticheatMgr.cpp:157]: Notes added
https://github.com/TrinityCore/TrinityCoreCustomChanges/pull/104/commits/e67649b95c0c29fea9a7767f4ff1b08d281c1a0f

**Pull request merged**